patches/server/1045-Add-player-kill-credit-to-beds-respawn-anchors.patch

```diff
@@ -0,0 +1,120 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: 2stinkysocks <zmehall@gmail.com>
+Date: Mon, 29 Jan 2024 00:30:34 -0700
+Subject: [PATCH] Add player kill credit to beds/respawn anchors
+
+
+diff --git a/src/main/java/net/minecraft/world/damagesource/DamageSource.java b/src/main/java/net/minecraft/world/damagesource/DamageSource.java
+index fc6903b20a6e084729306fc960a6fc80e094f76c..f6f1920d2fe3676b71fad187dfc488aa1689ed9f 100644
+--- a/src/main/java/net/minecraft/world/damagesource/DamageSource.java
++++ b/src/main/java/net/minecraft/world/damagesource/DamageSource.java
+@@ -81,6 +81,15 @@ public class DamageSource {
+         this(type, (Entity) null, (Entity) null, position);
+     }
+ 
++    // Paper start - add player kill credit
++    public DamageSource(Holder<DamageType> type, Vec3 position, @Nullable Entity attacker) {
++        this.type = type;
++        this.causingEntity = attacker;
```

Comment:
I am unsure how happy I am with hijacking this field.
Given neither paper nor spigot have a damage source API *yet* it might be fine, but this is changing vanilla behaviour and spigot behaviour.

Does anything stop us from going down the same path as the critical damage API and add a field + getter/setter for the "igniter" entity ref ?

Reply:
I think it makes more sense to just add a setter for the causingEntity field, since this is already used in LivingEntity's hurt method to determine kill credit, I can verify that this works later.

Reply:
I was more referring to the fact that other things might use the `causingEntity` field.
Changing that might break things we are not aware of and testing that everything still works fine is kind of hard.

Reply:
it's more the fact that "causingEntity" generally has a very explict meaning within the servers code and reusing this field is likely to cause some issues, if not now, maybe in the future; would much rather just add a field this rather than trying to hijack th emeaning of an exisitng one

Reply:
Wouldn't this just require more changes to `hurt()` in LivingEntity? It seems like all the existing logic in there would just have to be implemented twice, since the entity "causing" the event is still the person who ignited the block. I know that this avoids having unintended side effects with other parts of the code but I don't see a way it could be implemented cleanly without just mostly duplicating the logic.

Reply:
Well that is exactly the point tho. 

Vanilla does *not* consider the "igniting entity" the causing entity of the damage.
This PR should not change this. The logic in LivingEntity#hurt acts on the causingEntity of the damage source, which should remain null for explosions from e.g. respawn anchors.

You are merely supposed to hijack the DamageSource to pass through the entity that ignited the block so that the API event can expose it.
Changing what vanilla considers the actual causingEntity is not a good idea, which is why cat and I suggested this to be moved to its own field so existing logic does not change.

Reply:
I see how I was misunderstanding, it's definitely better for plugin developers to implement this themselves, sorry for the confusion. I'm starting to think this change would be better to make upstream, so I might create a similar pr to spigot that only changes the API. I'll close this pr and will come back to it if I run into problems with spigot.